### PR TITLE
Beta

### PR DESCRIPTION
**Pull request by wojtek-krysiak**
*

### CODE REVIEW

File: src/features/upload-file/factories/delete-file-factory.spec.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import { ActionRequest, RecordActionResponse, ActionContext, BaseRecord } from 'admin-bro'
```
Comment (wojtek-krysiak) [inline src/features/upload-file/factories/delete-file-factory.spec.ts:1]: to slim down main feature function I extracted all other functions creating hooks, etc to factoryFunction. Each of them has the same params: (UploadOptions and provider). So now I can easily test them.

File: src/features/upload-file/types/property-custom.type.ts [hunk 2]
```diff
@@ -7,14 +7,16 @@ import { MimeType } from './mime-types.type'
...
   filePathProperty: string,
   fileProperty: string,
+  filesToDeleteProperty: string,
   keyProperty: string,
   fileNameProperty?: string,
   mimeTypeProperty?: string,
   bucketProperty?: string,
   defaultBucket: string,
-  mimeTypes?: Array<MimeType>,
+  mimeTypes?: Array<MimeType | string>,
```
Comment (wojtek-krysiak) [inline src/features/upload-file/types/property-custom.type.ts:16]: this one caused problems when I used it outside this package - so I simply added string so tsc won complain when user gave some mimetype i don't have in list

File: src/features/upload-file/types/upload-options.type.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Property under which file name will be stored
+     */
+    filename?: string;
+  },
+  /**
+   * Function which defines where the file should be placed inside the bucket.
+   * Default to `${record.id()}/${filename}`.
+   */
+  uploadPath?: UploadPathFunction;
```
Comment (wojtek-krysiak) [inline src/features/upload-file/types/upload-options.type.ts:69]: this and the next one are new options

File: src/features/upload-file/upload-file.feature.ts [hunk 4]
```diff
@@ -1,175 +1,59 @@
...
-    if (key) {
-      // eslint-disable-next-line no-param-reassign
-      record.params[filePathProperty] = await provider.path(
-        key, storedBucket || provider.bucket, context,
-      )
-    }
-
-    return record
-  }
+  const stripFileFromPayload = stripPayloadFactory(configWithDefault)
```
Comment (wojtek-krysiak) [inline src/features/upload-file/upload-file.feature.ts:46]: as mentioned I refactored almost everything to these factories

File: example-app/src/admin/resources/create-resource-result.type.ts [hunk 5]
```diff
@@ -3,5 +3,5 @@ import { ResourceOptions, FeatureType } from 'admin-bro'
 export type CreateResourceResult<T> = {
   resource: T;
   options: ResourceOptions;
-  features: [FeatureType];
+  features: Array<FeatureType>;
```
Comment (plorencrstit) [inline example-app/src/admin/resources/create-resource-result.type.ts:6]: What is the purpose of this change? Is there any other difference than syntax?

**Review by plorencrstit — APPROVED** [review]

File: example-app/src/admin/resources/create-resource-result.type.ts (hunk 5 above)
Comment (wojtek-krysiak) [reply example-app/src/admin/resources/create-resource-result.type.ts:6]: no - they are different. `[FeaureType]` is a `Tuple` with the actual length, where `Array<FeatureType>` is an regular `Array`. The first one returns an error when you want to add more than one feature
